### PR TITLE
requirePeer

### PR DESCRIPTION
**Pull request by subtly**
Require a peer to be connected, regardless of node network.

### CODE REVIEW

**Comment by subtly** [comment]
#1416

File: libp2p/Host.cpp [hunk 1]
```diff
@@ -417,24 +399,58 @@ void Host::addNode(NodeId const& _node, std::string const& _addr, unsigned short
 		cwarn << "Private port being recorded - setting to 0";
 		_tcpPeerPort = 0;
 	}
+	
+	if (m_nodeTable) m_nodeTable->addNode(Node(_node, NodeIPEndpoint(bi::udp::endpoint(_addr, _udpNodePort), bi::tcp::endpoint(_addr, _tcpPeerPort))));
```
Comment (gavofyork) [inline libp2p/Host.cpp]: single-line if - breaks style

File: alethzero/MainWin.cpp [hunk 2]
```diff
@@ -251,7 +252,30 @@ void Main::addNewId(QString _ids)
 
 NetworkPreferences Main::netPrefs() const
 {
-	return NetworkPreferences(ui->port->value(), ui->forceAddress->text().toStdString(), ui->upnp->isChecked(), ui->localNetworking->isChecked());
+	auto listenip = ui->listenIP->text().toStdString();
```
Comment (gavofyork) [inline alethzero/MainWin.cpp]: should be `listenIP`, `publicIP`

File: alethzero/MainWin.cpp [hunk 3]
```diff
@@ -1765,13 +1785,26 @@ void Main::on_connect_triggered()
...
 	{
-		string host = s.section(":", 0, 0).toStdString();
-		unsigned short port = s.section(":", 1).toInt();
-		web3()->connect(host, port);
+		bool required = m_connect.required();
+		string host(m_connect.host().toUtf8().constData());
+		NodeId nodeid;
+		try
+		{
+			string nstr(m_connect.nodeId().toUtf8().constData());
```
Comment (gavofyork) [inline alethzero/MainWin.cpp]: `nstr`? something more descrptive. or at least pronounceable

File: alethzero/MainWin.cpp [hunk 4]
```diff
@@ -1765,13 +1785,26 @@ void Main::on_connect_triggered()
...
+			nodeid = NodeId(fromHex(nstr));
```
Comment (gavofyork) [inline alethzero/MainWin.cpp]: `nodeID`

Comment (gavofyork) [inline alethzero/MainWin.cpp]: `= NodeId(fromHex(m_connect.nodeId().toStdString()))` is cleaner and doesn't require the temp var.

File: alethzero/MainWin.cpp [hunk 5]
```diff
@@ -1765,13 +1785,26 @@ void Main::on_connect_triggered()
...
-	if (ok && s.contains(":"))
+	
+	m_connect.setEnvironment(m_servers);
+	if (m_connect.exec() == QDialog::Accepted)
 	{
-		string host = s.section(":", 0, 0).toStdString();
-		unsigned short port = s.section(":", 1).toInt();
-		web3()->connect(host, port);
+		bool required = m_connect.required();
+		string host(m_connect.host().toUtf8().constData());
```
Comment (gavofyork) [inline alethzero/MainWin.cpp]: yuck. why not `m_connect.host().toStdString()`?

File: alethzero/Connect.h [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+class Connect : public QDialog
+{
+    Q_OBJECT
+
+public:
+    explicit Connect(QWidget *parent = 0);
+    ~Connect();
+
+    void setEnvironment(QStringList const& _nodes);
```
Comment (gavofyork) [inline alethzero/Connect.h]: spaces?!

File: alethzero/Connect.h [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+	
+	void reset();
+
```
Comment (gavofyork) [inline alethzero/Connect.h]: an awful lot of blank lines in here

File: alethzero/Connect.h [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+namespace Ui { class Connect; }
+namespace dev { namespace p2p { class Host; } }
+
+class Connect : public QDialog
+{
+    Q_OBJECT
+
+public:
+    explicit Connect(QWidget *parent = 0);
```
Comment (gavofyork) [inline alethzero/Connect.h]: `QWidget* _parent`

File: alethzero/Connect.h [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+	void reset();
+	
+	QString host();
+	
+	QString nodeId();
+	
+	bool required();
+	
+private:
+    Ui::Connect *ui;
```
Comment (gavofyork) [inline alethzero/Connect.h]: i don't mind the `ui`, but should be `Ui::Connect* ui;`

File: libp2p/NodeTable.cpp [hunk 10]
```diff
@@ -550,31 +552,16 @@ void NodeTable::doRefreshBuckets(boost::system::error_code const& _ec)
...
-					{
-						refreshed = true;
-						ping(p.get());
-						break;
-					}
-					d.nodes.pop_front();
-				}
-			}
+		NodeId randNodeId;
+		crypto::Nonce::get().ref().copyTo(randNodeId.ref().cropped(0,h256::size));
```
Comment (gavofyork) [inline libp2p/NodeTable.cpp]: space after comma

File: libp2p/NodeTable.cpp [hunk 11]
```diff
@@ -550,31 +552,16 @@ void NodeTable::doRefreshBuckets(boost::system::error_code const& _ec)
...
+		crypto::Nonce::get().ref().copyTo(randNodeId.ref().cropped(h256::size,h256::size));
```
Comment (gavofyork) [inline libp2p/NodeTable.cpp]: same

**Comment by gavofyork** [comment]
rebase needed.

**Comment by gavofyork** [comment]
another rebase needed

File: alethzero/Connect.h [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+	void reset();
+
+	// Form field values:
+	
+	QString host();
+	QString nodeId();
+	bool required();
+
+private:
+    Ui::Connect* ui;
```
Comment (gavofyork) [inline alethzero/Connect.h]: spaces

File: alethzero/Connect.h [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+	explicit Connect(QWidget* _parent = 0);
+	~Connect();
+
+	void setEnvironment(QStringList const& _nodes);
+
+	/// clear dialogue inputs
+	void reset();
+
+	// Form field values:
+
```
Comment (gavofyork) [inline alethzero/Connect.h]: blank line?

File: alethzero/Connect.h [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+class Connect : public QDialog
+{
+	Q_OBJECT
+
+public:
+	explicit Connect(QWidget* _parent = 0);
+	~Connect();
+
+	void setEnvironment(QStringList const& _nodes);
```
Comment (gavofyork) [inline alethzero/Connect.h]: doc?

File: alethzero/MainWin.cpp [hunk 15]
```diff
@@ -251,7 +252,30 @@ void Main::addNewId(QString _ids)
...
+	}
+	
+	auto publicIP = ui->forcePublicIP->text().toStdString();
+	try
+	{
+		publicIP = bi::address::from_string(publicIP).to_string();
+	}
+	catch (...)
+	{
+		publicIP = "";
```
Comment (gavofyork) [inline alethzero/MainWin.cpp]: `.clear()` is better practice.

File: alethzero/MainWin.cpp [hunk 16]
```diff
@@ -251,7 +252,30 @@ void Main::addNewId(QString _ids)
...
 {
-	return NetworkPreferences(ui->port->value(), ui->forceAddress->text().toStdString(), ui->upnp->isChecked(), ui->localNetworking->isChecked());
+	auto listenIP = ui->listenIP->text().toStdString();
+	try
+	{
+		listenIP = bi::address::from_string(listenIP).to_string();
+	}
+	catch (...)
+	{
+		listenIP = "";
```
Comment (gavofyork) [inline alethzero/MainWin.cpp]: `.clear()`

File: alethzero/MainWin.cpp [hunk 17]
```diff
@@ -1765,13 +1785,25 @@ void Main::on_connect_triggered()
...
-		unsigned short port = s.section(":", 1).toInt();
-		web3()->connect(host, port);
+		bool required = m_connect.required();
+		string host(m_connect.host().toStdString());
+		NodeId nodeID;
+		try
+		{
+			nodeID = NodeId(fromHex(m_connect.nodeId().toStdString()));
+		}
+		catch (BadHexCharacter()) {}
```
Comment (gavofyork) [inline alethzero/MainWin.cpp]: i think you mean `BadHexCharacter` (no parens)

File: libp2p/Network.h [hunk 18]
```diff
@@ -39,12 +39,19 @@ namespace p2p
 
 struct NetworkPreferences
 {
-	NetworkPreferences(unsigned short p = 30303, std::string i = std::string(), bool u = true, bool l = false): listenPort(p), publicIP(i), upnp(u), localNetworking(l) {}
+	// Default Network Preferences
+	NetworkPreferences(unsigned short lp = 30303): listenPort(lp) {}
+	
+	// Network Preferences with specific Listen IP
+	NetworkPreferences(std::string l, unsigned short lp = 30303, bool u = true): publicIPAddress(), listenIPAddress(l), listenPort(lp), traverseNAT(u) {}
```
Comment (gavofyork) [inline libp2p/Network.h]: `std::string const&`?

File: libp2p/Network.h [hunk 19]
```diff
@@ -39,12 +39,19 @@ namespace p2p
...
+	
+	// Network Preferences with intended Public IP
+	NetworkPreferences(std::string publicIP, std::string l = std::string(), unsigned short lp = 30303, bool u = true): publicIPAddress(publicIP), listenIPAddress(l), listenPort(lp), traverseNAT(u) { if (!publicIPAddress.empty() && !isPublicAddress(publicIPAddress)) BOOST_THROW_EXCEPTION(InvalidPublicIPAddress()); }
```
Comment (gavofyork) [inline libp2p/Network.h]: `std::string const&`?

File: test/peer.cpp [hunk 20]
```diff
@@ -62,7 +62,7 @@ BOOST_AUTO_TEST_CASE(save_nodes)
 	std::list<Host*> hosts;
 	for (auto i:{0,1,2,3,4,5})
 	{
-		Host* h = new Host("Test", NetworkPreferences(30300 + i, "127.0.0.1", false, true));
+		Host* h = new Host("Test", NetworkPreferences("127.0.0.1", 30300 + i, false));
```
Comment (gavofyork) [inline test/peer.cpp]: probably not for this PR, but trying to evict all bool params these days. make an `enum class` for this bool param at some point.

File: alethzero/Connect.h [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+
+class Connect : public QDialog
+{
+	Q_OBJECT
+
+public:
+	explicit Connect(QWidget* _parent = 0);
+	~Connect();
+
+	/// Populate host chooser with default host entries
```
Comment (gavofyork) [inline alethzero/Connect.h]: Docs: all should be consistently capitalised and punctuated (i.e. real sentences!)

For UI stuff, describe in terms of user-expectation/meaning.

File: alethzero/Connect.h [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+	void setEnvironment(QStringList const& _nodes);
+
+	/// clear dialogue inputs
+	void reset();
+
+	/// @returns host string chosen or entered
+	QString host();
+	
+	/// @returns NodeId entered
```
Comment (gavofyork) [inline alethzero/Connect.h]: e.g. "@returns the identity of the node, as entered by the user. Assumed to be a 64-character hex string."

**Pull request merged**